hbase-server/src/main/java/org/apache/hadoop/hbase/master/webapp/MetaBrowser.java

```diff
@@ -0,0 +1,378 @@
+  }
+
+  private static TableName resolveScanTable(final HttpServletRequest request) {
+    final String requestValue = resolveRequestParameter(request, SCAN_TABLE_PARAM);
+    if (requestValue == null) {
+      return null;
+    }
+    return TableName.valueOf(requestValue);
+  }
+
+  private static String resolveRequestParameter(final HttpServletRequest request,
+    final String param) {
+    if (request == null) {
+      return null;
+    }
+    final String requestValueStrEnc = request.getParameter(param);
+    if (StringUtils.isBlank(requestValueStrEnc)) {
+      return null;
+    }
+    return urlDecode(requestValueStrEnc);
+  }
+
+  private static Filter buildTableFilter(final TableName tableName) {
+    return new PrefixFilter(tableName.toBytes());
+  }
+
+  private static Filter buildScanRegionStateFilter(final RegionState.State state) {
+    return new SingleColumnValueFilter(
+      HConstants.CATALOG_FAMILY,
+      HConstants.STATE_QUALIFIER,
+      CompareOperator.EQUAL,
+      // use the same serialization strategy as found in MetaTableAccessor#addRegionStateToPut
+      Bytes.toBytes(state.name()));
+  }
+
+  private Filter buildScanFilter() {
+    if (scanTable == null && scanRegionState == null) {
+      return null;
+    }
+
+    final List<Filter> filters = new ArrayList<>(2);
+    if (scanTable != null) {
+      filters.add(buildTableFilter(scanTable));
+    }
+    if (scanRegionState != null) {
+      filters.add(buildScanRegionStateFilter(scanRegionState));
+    }
+    if (filters.size() == 1) {
+      return filters.get(0);
+    }
+    return new FilterList(FilterList.Operator.MUST_PASS_ALL, filters);
+  }
+
+  private Scan buildScan() {
+    final Scan metaScan = new Scan()
+      .addFamily(HConstants.CATALOG_FAMILY)
+      .readVersions(1)
+      .setLimit((scanLimit != null ? scanLimit : SCAN_LIMIT_DEFAULT) + 1);
+    if (scanStart != null) {
+      metaScan.withStartRow(scanStart, false);
+    }
+    final Filter filter = buildScanFilter();
+    if (filter != null) {
+      metaScan.setFilter(filter);
+    }
+    return metaScan;
+  }
+
+  /**
+   * Adds {@code value} to {@code encoder} under {@code paramName} when {@code value} is non-null.
+   */
+  private <T> void addParam(final QueryStringEncoder encoder, final String paramName,
+    final T value) {
+    if (value != null) {
+      encoder.addParam(paramName, value.toString());
+    }
+  }
+
+  private QueryStringEncoder buildFirstPageEncoder() {
+    final QueryStringEncoder encoder =
+      new QueryStringEncoder(request.getRequestURI());
+    addParam(encoder, NAME_PARAM, name);
+    addParam(encoder, SCAN_LIMIT_PARAM, scanLimit);
+    addParam(encoder, SCAN_REGION_STATE_PARAM, scanRegionState);
+    addParam(encoder, SCAN_TABLE_PARAM, scanTable);
+    return encoder;
+  }
+
+  public String buildFirstPageUrl() {
+    return buildFirstPageEncoder().toString();
+  }
+
+  public static String buildStartParamFrom(final byte[] lastRow) {
```

Comment:
this can be package-protected as it's only used internally and within the unit test.